### ISSUE TITLE
Development

### ISSUE THREAD

**Issue by Fantu** [BetaCoinDev/betacoin#4]
Betacoin development currently seems null.
This git is not forked by bitcoin git but added in one commit after major of betacoin change was already added, so is long/difficult do a rebase for atleast add the improvements/fixes from bitcoin git.
I think that is good thing fork bitcoin git from latest stable version and add the commits about all betacoin changes.
So then it will be easier and faster to do a "git rebase" each new stable version of bitcoin for at least include the improvements/fixes made ​​on bitcoin git.

**Comment by Fantu** [comment]
To make a patch with all changes your made for betacoin should be this if I remember good:
Add bitcoin git as remote upstream and fetch it:
git remote add upstream https://github.com/bitcoin/bitcoin.git
git fetch upstream
Make all changes in unified diff:
git diff [hash of upstream commit where you start to fork] master > betacoin.patch
A good thing is to split the betacoin changes into small patch with description of changes to make easy resolve conflict on git rebase.

Waiting rebase with bitcoin upstream I think is important start to backport atleast security fixes missed:
https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures
Betacoin seems forked not from one stable version but from master during the ongoing development of 0.9.
Being put on github via a single commit with the changes already betacoin I can not figure out what commit was forked to understand what are the security commits missed that must be backported.

**Comment by Fantu** [comment]
@BetaCoinDev: If you write the commit where you start to fork I can try to make betacoin diff for you when I'll have time.

**Comment by Fantu** [comment]
Probably I found the commit of bitcoin when you started to fork:
dabd4bd48e3cc020d81baaa20dd03a1bf4972bec (Bugfix: Include block creation and SSL options headings in usage)
@BetaCoinDev: is correct?

Diff file from this and betacoin master is big, seems that there are unuseful file mode set to remove and some file rename not tracked, now I'll search a solution to minimize betacoin patches and track all.

**Comment by BetaCoinDev** [comment]
Oh, sorry for delay in answers i am really hard working on developing advanced betacoin blockchain explorer and e-commerce modules. 
There is no secret, i used Joulecoin source code - that is a fork of Bitcoin.

**Comment by Fantu** [comment]
Thanks for reply.
Fork without complete trace of all changes (betacoin) of one fork (joulecoin) based on other fork (zetacoin) based on one commit of bitcoin 0.9 (not a stable version but on development)... seems very bad :(
I'll try to do a rebase on bitcoin commit of start and keep trace of all changes to make easy future rebase of stable bitcoin version, I think that should a good idea, I saw some altcoins repository and I found only litecoin that seems to do a rebase on stable bitcoin versions and having an development active.

**Comment by BetaCoinDev** [comment]
Yep :) Thats not good, but i began that more as a research project - to make a "fair launch" of a new blockchain (i was experimenting with block time and difficulty retarget calculation a lot) and to apply inflation after main supply,   + to develop services for that project. And i made it in mind with block time (not low for higher confirmations value) and to have "main" coin units supply in circulation in as i think optimum time ~6 years. 
Fork from fork from fork :) is really not good but it works and it was approved on different blockchains with different parameters. Again, yes it is not good, and your work on rebasing it according to the stable and last Bitcoin release is really helpful, i really appreciate it!  
I am really glad that you are interested in Betacoin support and development, i appreciate any help.

**Comment by BetaCoinDev** [comment]
As for me, i plan to release updated Betacoin version in the beginning of January - so if you will make rebase based on last Bitcoin code - i will include that commits in new update and it will appear earlier.
Now i am working on blockchain explorer with advanced search and trying to implement in with visual graph support and maybe with some other really user friendly staff.
Also i am working on number of Betacoin e-commerce distributions to kickstart e-shops.
Thanks for your help & Happy Holidays!

**Comment by Fantu** [comment]
About full rebase on bitcoin upstream newer version will be good do it only with stable versions, starting from 0.9.
I'll start to do actual betacoin code with all bitcoin commits plus all betacoin patches with details to make easy and faster the future rebase.
There are the security commits of pull request and all future bitcoin security fixes that shold be good release as soon as possible without waiting the rebase.

**Comment by BetaCoinDev** [comment]
I added security fixes to master branch and asked pool owners to update the clients. Thank you for your support, please provide BET address for some tips.

**Comment by Fantu** [comment]
I did a draft with rebase on start commit of bitcoin and some improvements:
https://github.com/Fantu/betacoin/tree/rebase/staging

This patch need to be spitted and improved:
https://github.com/Fantu/betacoin/commit/43aba4e1d74270388725c1ce236dcdcf0ffa31f6
see details on commit descriptions.

About clean and improvement of generic code opinions are appreciated by any developer.

All other betacoin core changes must be splitted and well described by BetaCoinDev or other experienced developers about them.

Thanks for any reply and sorry for my bad english.

**Comment by Fantu** [comment]
I also saw that src/qt/res/icons/toolbar_testnet.png have joulecoin image instead of betacoin and some other images have size different from original.

Seems there is also exclusion of src/miner.h and src/miner.cpp and inclusion of they source on src/main.cpp.
If is modified is better maintain in original files for track the betacoin changed and reduce the patch size.

**Comment by Fantu** [comment]
I removed some unuseful things and I'll remove also changes to src/version.cpp.
Tomorrow probably I'll also remove duplicate functions into main.cpp to miner.cpp including betacoin changes, useful to decrase patch size and keep tracked betacoin changes.

On src/qt/intro.cpp, some lines removed near start of file are needed? If yes can you explain me because they have been removed?

**Comment by Fantu** [comment]
I did some other clean and small improvement on https://github.com/Fantu/betacoin/tree/rebase/staging, tested and build successfull.
On https://github.com/Fantu/betacoin/tree/rebase/devel I removed miner.h and miner.cpp exclusion from build, apply betacoin changes on them (seems that was very few) and remove duplicate code in main.h and main.cpp.

On build gave me these error:
_BUILD_INFO -DLINUX -D_FILE_OFFSET_BITS=64 -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/share/qt4/mkspecs/linux-g++-64 -I/usr/include/qt4/QtCore -I/usr/include/qt4/QtNetwork -I/usr/include/qt4/QtGui -I/usr/include/qt4 -Isrc -Isrc/json -Isrc/qt -Isrc/leveldb/include -Isrc/leveldb/helpers -Ibuild -Ibuild -o build/miner.o src/miner.cpp
src/miner.cpp: In function ‘CBlockTemplate\* CreateNewBlock(CReserveKey&)’:
src/miner.cpp:324:77: error: invalid initialization of reference of type ‘CCoinsViewCache&’ from expression of type ‘CValidationState’
src/main.h:313:6: error: in passing argument 2 of ‘bool UpdateCoins(const CTransaction&, CCoinsViewCache&, CTxUndo&, int, const uint256&)’
src/miner.cpp:363:83: error: ‘GetBlockValue’ was not declared in this scope
src/miner.cpp:369:72: error: ‘GetNextWorkRequired’ was not declared in this scope
src/miner.cpp: In function ‘void BitcoinMiner(CWallet_)’:
src/miner.cpp:565:40: error: ‘pwalletMain’ was not declared in this scope
make: *_\* [build/miner.o] Errore 1

From here I realized that the functions miner.cpp duplicate in main.cpp was only a workaround for these errors.

Someone can help me to improve the mining betacoin changes without using this bad workaround?

**Comment by Fantu** [comment]
I found the problem... joulecoin or zetacoin developer have reverted some patches and fixes without tracing them.
I not understand if they did it because was not working or because they did a wrong small rebase, with all altcoin changes mainly on one commit without good descriptions is difficult to understand :(

The best things should be know all really needed betacoin core changes and remove all others core changes that probably are only reverts of some bitcoin patches.

**Comment by Fantu** [comment]
I'm going crazy...
I'm checking the last bitcoin commits and I still have not found all the commits reversed.
I guess I'll make a bash script that controls the diff of the bitcoin commits before the fork and tells what were reversed and which are not, or we'll lose too much time.

UPDATE:
The bigger patch is not reversed, I reapply solving conflict with betacoin changes on main.cpp and build is working. The most probable case is that the programmer of zetacoin or joulecoin instead resolve conflicts has removed the changes to the files with conflict keeping the partial rebase.
I'll try to complete the rebase.

**Comment by Fantu** [comment]
Now the rebase with bitcoin commit of start seems complete and build test was successfull.
Details on description of latest commit:
https://github.com/Fantu/betacoin/commit/b869f4c37924a7af038353a54b2578c975e7e17a

**Comment by Fantu** [comment]
I found other patches not applied, in these days I'll try to recheck all code.

I saw that translations string have coin name changed on string inside core code and on translation files on both source and translation fields.
I think that is good not change coin name in core code and on source field but only on translation field of locale files to reduce by two thirds the changes for coin name change only and make rebase faster.
What do you think?

**Comment by Fantu** [comment]
I did many other improvements, now all I'll post all details here:
https://cryptocointalk.com/topic/2714-devel-improvements-on-actual-git-source/
